src/fparser/two/symbol_table.py

```diff
@@ -0,0 +1,402 @@
+#
+# Copyright (c) 2021, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+
+'''
+The fparser2 symbol-table module. Defines various classes as well as
+the single, global SYMBOL_TABLES instance. The latter is a container
+for all of the top-level scoping units encountered during parsing.
+
+'''
+
+from __future__ import absolute_import, print_function
+import six
+from collections import namedtuple
+
+
+class SymbolTableError(Exception):
+    ''' Base class exception for symbol-table related errors. '''
+
+
+class SymbolTables(object):
+    '''
+    Class encapsulating functionality for the global symbol-tables object.
+    This is a container for all symbol tables constructed while parsing
+    code. All names are converted to lower case (since Fortran is not
+    case sensitive).
+
+    '''
+    def __init__(self):
+        self._symbol_tables = {}
+        # Those classes that correspond to a new scoping unit
+        self._scoping_unit_classes = []
+        # The stack of symbol tables accessible in the current scope
+        self._scope_stack = []
+
+    def __str__(self):
+        result = ("SymbolTables: {0} tables\n"
+                  "========================\n".format(
+                      len(self._symbol_tables)))
+        return result + "\n".join(sorted(self._symbol_tables.keys()))
+
+    def clear(self):
```

Comment:
Is this only expected to be used in tests? If so, perhaps this method should not exist.

Reply:
I didn't imagine it being used only for tests. It may be necessary e.g. for tutorials as otherwise you've got no way of clearing the symbol tables if you then want to parse a new bit of code.